src/hotspot/share/opto/superword.cpp

```diff
@@ -2975,51 +2787,64 @@ bool SuperWord::is_vector_use(Node* use, int u_idx) const {
     return true;
   }
 
+  if (!is_velt_basic_type_compatible_use_def(use, u_idx)) {
+    return false;
+  }
+
   if (VectorNode::is_muladds2i(use)) {
-    // MulAddS2I takes shorts and produces ints - hence the special checks
-    // on alignment and size.
+    // MulAddS2I takes shorts and produces ints.
     if (u_pk->size() * 2 != d_pk->size()) {
       return false;
     }
-    for (uint i = 0; i < MIN2(d_pk->size(), u_pk->size()); i++) {
-      Node* ui = u_pk->at(i);
-      Node* di = d_pk->at(i);
-      if (alignment(ui) != alignment(di) * 2) {
-        return false;
-      }
-    }
     return true;
   }
 
-  if (u_pk->size() != d_pk->size())
+  if (u_pk->size() != d_pk->size()) {
     return false;
-
-  if (longer_type_for_conversion(use) != T_ILLEGAL) {
-    // These opcodes take a type of a kind of size and produce a type of
-    // another size - hence the special checks on alignment and size.
-    for (uint i = 0; i < u_pk->size(); i++) {
-      Node* ui = u_pk->at(i);
-      Node* di = d_pk->at(i);
-      if (ui->in(u_idx) != di) {
-        return false;
-      }
-      if (alignment(ui) / type2aelembytes(velt_basic_type(ui)) !=
-          alignment(di) / type2aelembytes(velt_basic_type(di))) {
```

Comment:
Note: we scaled the alignment by the element size. This allows us the transitions when doing type conversion, i.e. from 4 bytes to 8 bytes.